channeldb/graphsession/graph_session.go

```diff
@@ -0,0 +1,141 @@
+package graphsession
+
+import (
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/kvdb"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing"
+	"github.com/lightningnetwork/lnd/routing/route"
+)
+
+// Factory implements the routing.GraphSessionFactory and can be used to start
+// a session with a ReadOnlyGraph.
+type Factory struct {
+	graph ReadOnlyGraph
+}
+
+// NewFactory constructs a new Factory which can then be used to start a new
+// Session.
+func NewFactory(graph ReadOnlyGraph) routing.GraphSessionFactory {
+	return &Factory{
+		graph: graph,
+	}
+}
+
+// NewSession will produce a new GraphSession to use for a path-finding session.
+// The caller MUST call Close once it has finished using the session.
+//
+// This is part of the routing.GraphSessionFactory interface.
+func (g *Factory) NewSession() (routing.GraphSession, error) {
+	tx, err := g.graph.NewPathFindTx()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Session{
+		graph: g.graph,
+		tx:    tx,
+	}, nil
+}
+
+// A compile-time check to ensure that Factory implements the
+// routing.GraphSessionFactory interface.
+var _ routing.GraphSessionFactory = (*Factory)(nil)
+
+// Session is an implementation of the routing.GraphSession interface and can
+// be used to access the backing channel graph.
+type Session struct {
+	graph graph
+	tx    kvdb.RTx
+}
+
+// NewRoutingGraph constructs a Session that which does not first start a
```

Comment:
meaning each call to the routing.Graph would start its own tx. Compared to the FactorySession which will start a read tx and pass the same same one into each call and only commits on Close.

Reply:
Since this relies on the behavior of the underlying implementation (`channeldb.ChannelGraph`) to create a new transaction if we don't pass one in, maybe we should make that assumption more clear on the `ForEachNodeDirectedChannel` interface definition?